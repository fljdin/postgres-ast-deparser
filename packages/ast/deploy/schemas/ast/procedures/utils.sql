-- Deploy schemas/ast/procedures/utils to pg

-- requires: schemas/ast/schema

BEGIN;


COMMIT;
