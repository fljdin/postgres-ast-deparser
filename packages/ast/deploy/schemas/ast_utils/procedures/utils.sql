-- Deploy schemas/ast_utils/procedures/utils to pg

-- requires: schemas/ast_utils/schema

BEGIN;

CREATE FUNCTION ast_utils.interval (
  n int
) returns text[] as $$
	select (CASE 
WHEN ( n = 2 ) THEN ARRAY[ 'month' ]
WHEN ( n = 4 ) THEN ARRAY[ 'year' ]
WHEN ( n = 6 ) THEN ARRAY[ 'year', 'month' ]
WHEN ( n = 8 ) THEN ARRAY[ 'day' ]
WHEN ( n = 1024 ) THEN ARRAY[ 'hour' ]
WHEN ( n = 1032 ) THEN ARRAY[ 'day', 'hour' ]
WHEN ( n = 2048 ) THEN ARRAY[ 'minute' ]
WHEN ( n = 3072 ) THEN ARRAY[ 'hour', 'minute' ]
WHEN ( n = 3080 ) THEN ARRAY[ 'day', 'minute' ]
WHEN ( n = 4096 ) THEN ARRAY[ 'second' ]
WHEN ( n = 6144 ) THEN ARRAY[ 'minute', 'second' ]
WHEN ( n = 7168 ) THEN ARRAY[ 'hour', 'second' ]
WHEN ( n = 7176 ) THEN ARRAY[ 'day', 'second' ]
WHEN ( n = 32767 ) THEN ARRAY[]::text[]
END);
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.reserved (
  str text
) returns boolean as $$
	select exists( select 1 from pg_get_keywords() where catcode = 'R' AND word=str  );
$$  
LANGUAGE 'sql' SECURITY DEFINER;

CREATE FUNCTION ast_utils.objtypes ()
returns text[] as $$
	select ARRAY[ 'ACCESS METHOD', 'AGGREGATE', NULL, NULL, NULL, 'CAST', 'COLUMN', 'COLLATION', 'CONVERSION', 'DATABASE', NULL, NULL, 'DOMAIN', 'CONSTRAINT', NULL, 'EXTENSION', 'FOREIGN DATA WRAPPER', 'SERVER', 'FOREIGN TABLE', 'FUNCTION', 'INDEX', 'LANGUAGE', 'LARGE OBJECT', 'MATERIALIZED VIEW', 'OPERATOR CLASS', 'OPERATOR', 'OPERATOR FAMILY', 'POLICY', NULL, NULL, 'ROLE', 'RULE', 'SCHEMA', 'SEQUENCE', NULL, 'STATISTICS', 'CONSTRAINT', 'TABLE', 'TABLESPACE', 'TRANSFORM', 'TRIGGER', 'TEXT SEARCH CONFIGURATION', 'TEXT SEARCH DICTIONARY', 'TEXT SEARCH PARSER', 'TEXT SEARCH TEMPLATE', 'TYPE', NULL, 'VIEW' ]::text[];
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.constrainttype_idxs (typ text)
returns int as $$
  select (CASE
WHEN (typ = 'CONSTR_NULL') THEN 0
WHEN (typ = 'CONSTR_NOTNULL') THEN 1
WHEN (typ = 'CONSTR_DEFAULT') THEN 2
WHEN (typ = 'CONSTR_IDENTITY') THEN 3
WHEN (typ = 'CONSTR_CHECK') THEN 4
WHEN (typ = 'CONSTR_PRIMARY') THEN 5
WHEN (typ = 'CONSTR_UNIQUE') THEN 6
WHEN (typ = 'CONSTR_EXCLUSION') THEN 7
WHEN (typ = 'CONSTR_FOREIGN') THEN 8
WHEN (typ = 'CONSTR_ATTR_DEFERRABLE') THEN 9
WHEN (typ = 'CONSTR_ATTR_NOT_DEFERRABLE') THEN 10
WHEN (typ = 'CONSTR_ATTR_DEFERRED') THEN 11
WHEN (typ = 'CONSTR_ATTR_IMMEDIATE') THEN 12
END);
$$
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.constrainttypes (contype int)
returns text as $$
  select (CASE
WHEN (contype =  0 ) THEN 'NULL'
WHEN (contype =  1 ) THEN 'NOT NULL'
WHEN (contype =  2 ) THEN 'DEFAULT'
WHEN (contype =  4 ) THEN 'CHECK'
WHEN (contype =  5 ) THEN 'PRIMARY KEY'
WHEN (contype =  6 ) THEN 'UNIQUE'
WHEN (contype =  7 ) THEN 'EXCLUDE'
WHEN (contype =  8 ) THEN 'REFERENCES'
END);
$$
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.objtypes_idxs (typ text)
returns int as $$
	select (CASE
WHEN (typ = 'OBJECT_ACCESS_METHOD') THEN 0
WHEN (typ = 'OBJECT_AGGREGATE') THEN 1
WHEN (typ = 'OBJECT_AMOP') THEN 2
WHEN (typ = 'OBJECT_AMPROC') THEN 3
WHEN (typ = 'OBJECT_ATTRIBUTE') THEN 4
WHEN (typ = 'OBJECT_CAST') THEN 5
WHEN (typ = 'OBJECT_COLUMN') THEN 6
WHEN (typ = 'OBJECT_COLLATION') THEN 7
WHEN (typ = 'OBJECT_CONVERSION') THEN 8
WHEN (typ = 'OBJECT_DATABASE') THEN 9
WHEN (typ = 'OBJECT_DEFAULT') THEN 10
WHEN (typ = 'OBJECT_DEFACL') THEN 11
WHEN (typ = 'OBJECT_DOMAIN') THEN 12
WHEN (typ = 'OBJECT_DOMCONSTRAINT') THEN 13
WHEN (typ = 'OBJECT_EVENT_TRIGGER') THEN 14
WHEN (typ = 'OBJECT_EXTENSION') THEN 15
WHEN (typ = 'OBJECT_FDW') THEN 16
WHEN (typ = 'OBJECT_FOREIGN_SERVER') THEN 17
WHEN (typ = 'OBJECT_FOREIGN_TABLE') THEN 18
WHEN (typ = 'OBJECT_FUNCTION') THEN 19
WHEN (typ = 'OBJECT_INDEX') THEN 20
WHEN (typ = 'OBJECT_LANGUAGE') THEN 21
WHEN (typ = 'OBJECT_LARGEOBJECT') THEN 22
WHEN (typ = 'OBJECT_MATVIEW') THEN 23
WHEN (typ = 'OBJECT_OPCLASS') THEN 24
WHEN (typ = 'OBJECT_OPERATOR') THEN 25
WHEN (typ = 'OBJECT_OPFAMILY') THEN 26
WHEN (typ = 'OBJECT_POLICY') THEN 27
WHEN (typ = 'OBJECT_PUBLICATION') THEN 28
WHEN (typ = 'OBJECT_PUBLICATION_REL') THEN 29
WHEN (typ = 'OBJECT_ROLE') THEN 30
WHEN (typ = 'OBJECT_RULE') THEN 31
WHEN (typ = 'OBJECT_SCHEMA') THEN 32
WHEN (typ = 'OBJECT_SEQUENCE') THEN 33
WHEN (typ = 'OBJECT_SUBSCRIPTION') THEN 34
WHEN (typ = 'OBJECT_STATISTIC_EXT') THEN 35
WHEN (typ = 'OBJECT_TABCONSTRAINT') THEN 36
WHEN (typ = 'OBJECT_TABLE') THEN 37
WHEN (typ = 'OBJECT_TABLESPACE') THEN 38
WHEN (typ = 'OBJECT_TRANSFORM') THEN 39
WHEN (typ = 'OBJECT_TRIGGER') THEN 40
WHEN (typ = 'OBJECT_TSCONFIGURATION') THEN 41
WHEN (typ = 'OBJECT_TSDICTIONARY') THEN 42
WHEN (typ = 'OBJECT_TSPARSER') THEN 43
WHEN (typ = 'OBJECT_TSTEMPLATE') THEN 44
WHEN (typ = 'OBJECT_TYPE') THEN 45
WHEN (typ = 'OBJECT_USER_MAPPING') THEN 46
WHEN (typ = 'OBJECT_VIEW') THEN 47
END);
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.getgrantobject (node jsonb)
returns text as $$
DECLARE
  objtype int;
  targtype int;
BEGIN 

  objtype = (node->'objtype')::int;
  IF (node->'targtype') IS NOT NULL THEN
    targtype = (node->'targtype')::int;
  END IF;

  IF (objtype = 0) THEN
    RETURN 'COLUMN';
  ELSIF (objtype = 1) THEN
    IF (targtype = 1) THEN 
      RETURN 'ALL TABLES IN SCHEMA';
    ELSIF (targtype = 2) THEN 
      RETURN 'TABLES';
    END IF;
    -- TODO could be a view
    RETURN 'TABLE';
  ELSIF (objtype = 2) THEN
    RETURN 'SEQUENCE';
  ELSIF (objtype = 3) THEN
    RETURN 'DATABASE';
  ELSIF (objtype = 4) THEN
    RETURN 'DOMAIN';
  ELSIF (objtype = 5) THEN
    RETURN 'FOREIGN DATA WRAPPER';
  ELSIF (objtype = 6) THEN
    RETURN 'FOREIGN SERVER';
  ELSIF (objtype = 7) THEN
    IF (targtype = 1) THEN 
      RETURN 'ALL FUNCTIONS IN SCHEMA';
    ELSIF (targtype = 2) THEN 
      RETURN 'FUNCTIONS';
    END IF;
    RETURN 'FUNCTION';
  ELSIF (objtype = 8) THEN
    RETURN 'LANGUAGE';
  ELSIF (objtype = 9) THEN
    RETURN 'LARGE OBJECT';
  ELSIF (objtype = 10) THEN
    RETURN 'SCHEMA';
  ELSIF (objtype = 11) THEN
    RETURN 'TABLESPACE';
  ELSIF (objtype = 12) THEN
    RETURN 'TYPE';
  END IF;

  RAISE EXCEPTION 'BAD_EXPRESSION %', 'GrantObjectType';

END;
$$  
LANGUAGE 'plpgsql' IMMUTABLE;

COMMIT;
