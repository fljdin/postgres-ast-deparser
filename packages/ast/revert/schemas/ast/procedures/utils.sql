-- Revert schemas/ast/procedures/utils from pg

BEGIN;

DROP FUNCTION ast.utils;

COMMIT;
