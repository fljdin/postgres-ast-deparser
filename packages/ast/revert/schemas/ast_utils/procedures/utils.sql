-- Revert schemas/ast_utils/procedures/utils from pg

BEGIN;

DROP FUNCTION ast_utils.utils;

COMMIT;
