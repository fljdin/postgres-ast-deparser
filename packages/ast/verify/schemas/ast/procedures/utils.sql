-- Verify schemas/ast/procedures/utils  on pg

BEGIN;

SELECT verify_function ('ast.utils');

ROLLBACK;
