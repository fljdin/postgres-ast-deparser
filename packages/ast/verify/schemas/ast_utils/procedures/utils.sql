-- Verify schemas/ast_utils/procedures/utils  on pg

BEGIN;

SELECT verify_function ('ast_utils.utils');

ROLLBACK;
