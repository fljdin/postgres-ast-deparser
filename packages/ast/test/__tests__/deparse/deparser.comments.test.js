import { getConnections } from '../../utils';

let db, teardown;
const objs = {
  tables: {}
};

beforeAll(async () => {
  ({ db, teardown } = await getConnections());
  await db.begin();
  await db.savepoint();
});
afterAll(async () => {
  try {
    //try catch here allows us to see the sql parsing issues!
    await db.rollback();
    await db.commit();
    await teardown();
  } catch (e) {
    // noop
  }
});

// WE NEED TO COMMENT OUT or whitelist m2m relations...

it('OBJECT_FUNCTION comment', async () => {
  const [{ expression: result }] = await db.any(`select deparser.expression(
    ast.comment_stmt(
        v_objtype := ast_constants.object_type('OBJECT_FUNCTION'),
        v_object := ast.object_with_args(
            v_objname := to_jsonb(ARRAY[
                ast.string('my_function')
            ]),
            v_objargs := to_jsonb(ARRAY[
                ast.type_name(
                    v_names := to_jsonb(ARRAY[ 
                        ast.string('arg_type_1')
                    ])
                ),
                ast.type_name(
                    v_names := to_jsonb(ARRAY[ 
                        ast.string('arg_type_2')
                    ])
                )
            ])
        ),
        v_comment := 'asdf'
    )
);`);
  expect(result).toMatchSnapshot();
});

it('helpers function', async () => {
  const [{ expression: result }] = await db.any(`select deparser.expression(
      ast_helpers.set_comment_on_function(
          v_schema_name := 'schema-name',
          v_function_name := 'function-name',
          v_comment := 'hithere'
      )
  );`);
  expect(result).toMatchSnapshot();
});

it('smart_comments helpers function', async () => {
  const [{ expression: result }] = await db.any(
    `select deparser.expression(
      ast_helpers.set_comment_on_function(
          v_schema_name := 'schema-name',
          v_function_name := 'function-name',
          v_tags := $1::jsonb,
          v_description := $2
      )
  );`,
    [{ b: 1, c: [0, 1, 2, 3] }, 'my description']
  );
  expect(result).toMatchSnapshot();
});
it('helpers table', async () => {
  const [{ expression: result }] = await db.any(`select deparser.expression(
    ast_helpers.set_comment(
          v_objtype := ast_constants.object_type('OBJECT_TABLE'),
          v_comment := 'hithere',
          variadic v_name := ARRAY[
              'schema-name',
              'table-name'
          ]
      )
  );`);
  expect(result).toMatchSnapshot();
});

it('smart_comments helpers table', async () => {
  const [{ expression: result }] = await db.any(
    `select deparser.expression(
    ast_helpers.set_comment(
          v_objtype := ast_constants.object_type('OBJECT_TABLE'),
          v_tags := $1::jsonb,
          v_description := $2,
          variadic v_name := ARRAY[
              'schema_name',
              'table_name'
          ]
      )
  );`,
    [
      {
        here: 'wego'
      },
      'my description'
    ]
  );
  expect(result).toMatchSnapshot();
});

it('smart_comments helpers table', async () => {
  const [{ expression: result }] = await db.any(
    `select deparser.expression(
    ast_helpers.set_comment(
          v_objtype := ast_constants.object_type('OBJECT_TABLE'),
          v_tags := $1::jsonb,
          v_description := $2,
          variadic v_name := ARRAY[
              'schema_name',
              'table_name'
          ]
      )
  );`,
    [
      {
        here: 'wego'
      },
      null
    ]
  );
  expect(result).toMatchSnapshot();
});

it('smart_comments helpers table', async () => {
  const [{ expression: result }] = await db.any(
    `select deparser.expression(
    ast_helpers.set_comment(
          v_objtype := ast_constants.object_type('OBJECT_TABLE'),
          v_tags := $1::jsonb,
          v_description := $2,
          variadic v_name := ARRAY[
              'schema_name',
              'table_name'
          ]
      )
  );`,
    [null, 'my descript']
  );
  expect(result).toMatchSnapshot();
});

it('smart_comments helpers table', async () => {
  const [{ expression: result }] = await db.any(
    `select deparser.expression(
    ast_helpers.set_comment(
          v_objtype := ast_constants.object_type('OBJECT_TABLE'),
          v_tags := $1::jsonb,
          v_description := $2,
          variadic v_name := ARRAY[
              'schema_name',
              'table_name'
          ]
      )
  );`,
    [null, null]
  );
  expect(result).toMatchSnapshot();
});

it('smart_comments helpers column', async () => {
  const [{ expression: result }] = await db.any(
    `select deparser.expression(
    ast_helpers.set_comment(
          v_objtype := ast_constants.object_type('OBJECT_COLUMN'),
          v_tags := $1::jsonb,
          v_description := $2,
          variadic v_name := ARRAY[
              'schema_name',
              'table_name',
              'column_name'
          ]
      )
  );`,
    [
      {
        here: 'wego'
      },
      'my description'
    ]
  );
  expect(result).toMatchSnapshot();
});

it('helpers table constraint', async () => {
  const [{ expression: result }] = await db.any(`select deparser.expression(
    ast_helpers.set_comment(
          v_objtype := ast_constants.object_type('OBJECT_TABCONSTRAINT'),
          v_comment := 'hithere',
          variadic v_name := ARRAY[
              'schema-name',
              'table_name',
              'my_constraint'
          ]
      )
  );`);
  expect(result).toMatchSnapshot();
});

it('helpers table constraint', async () => {
  const [{ expression: result }] = await db.any(`select deparser.expression(
    ast_helpers.set_comment(
          v_objtype := ast_constants.object_type('OBJECT_TABCONSTRAINT'),
          v_comment := 'hithere',
          variadic v_name := ARRAY[
              'table_name',
              'my_constraint'
          ]
      )
  );`);
  expect(result).toMatchSnapshot();
});

it('OBJECT_COLUMN comment', async () => {
  const [{ expression: result }] = await db.any(`select deparser.expression(
    ast.comment_stmt(
        v_objtype := ast_constants.object_type('OBJECT_COLUMN'),
        v_object := to_jsonb(ARRAY[
            ast.string('my_schema'),
            ast.string('my_table'),
            ast.string('my_column')
        ]),
        v_comment := 'asdf'
    )
);`);
  expect(result).toMatchSnapshot();
});

it('OBJECT_TABLE comment', async () => {
  const [{ expression: result }] = await db.any(`select deparser.expression(
    ast.comment_stmt(
        v_objtype := ast_constants.object_type('OBJECT_TABLE'),
        v_object := to_jsonb(ARRAY[
            ast.string('my_schema'),
            ast.string('my_table')
        ]),
        v_comment := 'asdf'
    )
);`);
  expect(result).toMatchSnapshot();
});

it('OBJECT_VIEW comment', async () => {
  const [{ expression: result }] = await db.any(`select deparser.expression(
    ast.comment_stmt(
        v_objtype := ast_constants.object_type('OBJECT_VIEW'),
        v_object := to_jsonb(ARRAY[
            ast.string('my_schema'),
            ast.string('my_view')
        ]),
        v_comment := 'asdf'
    )
);`);
  expect(result).toMatchSnapshot();
});

it('OBJECT_MATVIEW comment', async () => {
  const [{ expression: result }] = await db.any(`select deparser.expression(
    ast.comment_stmt(
        v_objtype := ast_constants.object_type('OBJECT_MATVIEW'),
        v_object := to_jsonb(ARRAY[
            ast.string('my_schema'),
            ast.string('my_view')
        ]),
        v_comment := 'asdf'
    )
);`);
  expect(result).toMatchSnapshot();
});

it('OBJECT_TYPE comment', async () => {
  const [{ expression: result }] = await db.any(`select deparser.expression(
    ast.comment_stmt(
        v_objtype := ast_constants.object_type('OBJECT_TYPE'),
        v_object := to_jsonb(ARRAY[
            ast.string('my_schema'),
            ast.string('my_type')
        ]),
        v_comment := 'asdf'
    )
);`);
  expect(result).toMatchSnapshot();
});

it('OBJECT_TABCONSTRAINT comment', async () => {
  const [{ expression: result }] = await db.any(`select deparser.expression(
    ast.comment_stmt(
        v_objtype := ast_constants.object_type('OBJECT_TABCONSTRAINT'),
        v_object := to_jsonb(ARRAY[
            ast.string('my_schema'),
            ast.string('my_table'),
            ast.string('my_constraint')
        ]),
        v_comment := 'asdf'
    )
);`);
  expect(result).toMatchSnapshot();
});

it('smart_comments for any object', async () => {
  const [{ expression: result }] = await db.any(
    `select deparser.expression(
    ast.comment_stmt(
        v_objtype := ast_constants.object_type('OBJECT_TABCONSTRAINT'),
        v_object := to_jsonb(ARRAY[
            ast.string('my_schema'),
            ast.string('my_table'),
            ast.string('my_constraint')
        ]),
        v_comment := ast_helpers.smart_comments(
            tags := $1::jsonb,
            description := $2
        )
    )
);`,
    [{ a: 1 }, 'my description']
  );
  expect(result).toMatchSnapshot();
});

it('smart_comments for any complext', async () => {
  const [{ expression: result }] = await db.any(
    `select deparser.expression(
    ast.comment_stmt(
        v_objtype := ast_constants.object_type('OBJECT_TABCONSTRAINT'),
        v_object := to_jsonb(ARRAY[
            ast.string('my_schema'),
            ast.string('my_table'),
            ast.string('my_constraint')
        ]),
        v_comment := ast_helpers.smart_comments(
            tags := $1::jsonb,
            description := $2
        )
    )
);`,
    [
      {
        type: 'object',
        properties: {
          bookmarks: {
            type: 'array',
            minItems: 3,
            maxItems: 5,
            items: {
              type: 'object',
              properties: {
                id: {
                  type: 'number',
                  unique: true,
                  minimum: 1
                },
                url: {
                  type: 'string',
                  faker: 'internet.url'
                },
                title: {
                  type: 'string'
                },
                tags: {
                  type: 'string',
                  faker: 'custom.tags'
                }
              },
              required: ['id', 'url', 'title', 'tags']
            }
          }
        },
        required: ['bookmarks']
      },
      'my description'
    ]
  );
  expect(result).toMatchSnapshot();
});
