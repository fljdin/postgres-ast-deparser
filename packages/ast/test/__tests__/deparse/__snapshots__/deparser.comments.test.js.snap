// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OBJECT_COLUMN comment 1`] = `"COMMENT ON COLUMN my_schema.my_table.my_column IS 'asdf'"`;

exports[`OBJECT_FUNCTION comment 1`] = `"COMMENT ON FUNCTION my_function ( arg_type_1, arg_type_2 ) IS 'asdf'"`;

exports[`OBJECT_MATVIEW comment 1`] = `"COMMENT ON MATERIALIZED VIEW my_schema.my_view IS 'asdf'"`;

exports[`OBJECT_TABCONSTRAINT comment 1`] = `"COMMENT ON CONSTRAINT my_constraint ON my_schema.my_table IS 'asdf'"`;

exports[`OBJECT_TABLE comment 1`] = `"COMMENT ON TABLE my_schema.my_table IS 'asdf'"`;

exports[`OBJECT_TYPE comment 1`] = `"COMMENT ON TYPE my_schema.my_type IS 'asdf'"`;

exports[`OBJECT_VIEW comment 1`] = `"COMMENT ON VIEW my_schema.my_view IS 'asdf'"`;

exports[`helpers function 1`] = `"COMMENT ON FUNCTION \\"schema-name\\".\\"function-name\\" IS 'hithere';"`;

exports[`helpers table 1`] = `"COMMENT ON TABLE \\"schema-name\\".\\"table-name\\" IS 'hithere';"`;

exports[`helpers table constraint 1`] = `"COMMENT ON CONSTRAINT my_constraint ON \\"schema-name\\".table_name IS 'hithere';"`;

exports[`helpers table constraint 2`] = `"COMMENT ON CONSTRAINT my_constraint ON table_name IS 'hithere';"`;

exports[`smart_comments for any complext 1`] = `"COMMENT ON CONSTRAINT my_constraint ON my_schema.my_table IS E'@type object\\\\n@required bookmarks\\\\n@properties {\\"bookmarks\\": {\\"type\\": \\"array\\", \\"items\\": {\\"type\\": \\"object\\", \\"required\\": [\\"id\\", \\"url\\", \\"title\\", \\"tags\\"], \\"properties\\": {\\"id\\": {\\"type\\": \\"number\\", \\"unique\\": true, \\"minimum\\": 1}, \\"url\\": {\\"type\\": \\"string\\", \\"faker\\": \\"internet.url\\"}, \\"tags\\": {\\"type\\": \\"string\\", \\"faker\\": \\"custom.tags\\"}, \\"title\\": {\\"type\\": \\"string\\"}}}, \\"maxItems\\": 5, \\"minItems\\": 3}}\\\\nmy description'"`;

exports[`smart_comments for any object 1`] = `"COMMENT ON CONSTRAINT my_constraint ON my_schema.my_table IS E'@a 1\\\\nmy description'"`;

exports[`smart_comments helpers column 1`] = `"COMMENT ON COLUMN schema_name.table_name.column_name IS E'@here wego\\\\nmy description';"`;

exports[`smart_comments helpers function 1`] = `"COMMENT ON FUNCTION \\"schema-name\\".\\"function-name\\" IS E'@b 1\\\\n@c 0\\\\n@c 1\\\\n@c 2\\\\n@c 3\\\\nmy description';"`;

exports[`smart_comments helpers table 1`] = `"COMMENT ON TABLE schema_name.table_name IS E'@here wego\\\\nmy description';"`;

exports[`smart_comments helpers table 2`] = `"COMMENT ON TABLE schema_name.table_name IS E'@here wego';"`;

exports[`smart_comments helpers table 3`] = `"COMMENT ON TABLE schema_name.table_name IS E'my descript';"`;

exports[`smart_comments helpers table 4`] = `"COMMENT ON TABLE schema_name.table_name IS NULL;"`;
